src/Collections/Arrays/BinarySearch.dfy

```diff
@@ -0,0 +1,135 @@
+  {
+    || n == |x|
+    || (n != |y| && x[n] < y[n])
+    || (n != |y| && x[n] == y[n] && LexicographicByteSeqBelowAux(x, y, n + 1))
+  }
+
+  lemma AboutLexicographicByteSeqBelow()
+    ensures TotalOrdering(LexicographicByteSeqBelow)
+  {
+    assert Reflexive(LexicographicByteSeqBelow) by {
+      forall x, n | 0 <= n <= |x| {
+        AboutLexicographicByteSeqBelowAux_Reflexive(x, n);
+      }
+    }
+    assert AntiSymmetric(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat |
+        n <= |x| && n <= |y| && x[..n] == y[..n] &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+      {
+        AboutLexicographicByteSeqBelowAux_AntiSymmetric(x, y, n);
+      }
+    }
+    assert StandardLibrary.Transitive(LexicographicByteSeqBelow) by {
+      forall x, y, z, n: nat |
+        n <= |x| && n <= |y| && n <= |z| &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+      {
+        AboutLexicographicByteSeqBelowAux_Transitive(x, y, z, n);
+      }
+    }
+    assert Connected(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat | n <= |x| && n <= |y| {
+        AboutLexicographicByteSeqBelowAux_Connected(x, y, n);
+      }
+    }
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Reflexive(x: seq<uint8>, n: nat)
+    requires n <= |x|
+    ensures LexicographicByteSeqBelowAux(x, x, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_AntiSymmetric(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    requires x[..n] == y[..n]
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+    ensures x == y
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Transitive(x: seq<uint8>, y: seq<uint8>, z: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y| && n <= |z|
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+    ensures LexicographicByteSeqBelowAux(x, z, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Connected(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    ensures LexicographicByteSeqBelowAux(x, y, n) || LexicographicByteSeqBelowAux(y, x, n)
+    decreases |x| - n
+  {
+  }
+
+method BinarySearch<T>(a: array<T>, key: T,compare: (T, T) -> bool) returns (r : nat)
+  requires forall i,j :: 0 <= i < j < a.Length ==> compare(a[i], a[j])
+  ensures 0 <= r ==> r < a.Length && a[r] == key
+  ensures r < 0 ==> key !in a[..]
+  requires StandardLibrary.Transitive(compare)
+  requires Connected(compare)
+{
+ var lo, hi : nat := 0, a.Length;
+  while lo < hi
+    invariant 0 <= lo <= hi <= a.Length
+    invariant key !in a[..lo] && key !in a[hi..]
+  {
+    var mid := (lo + hi) / 2;
+    if compare(key , a[mid]) {
+      hi := mid;
+    } else if compare(a[mid] , key) {
+      lo := mid + 1;
+    } else {
+      return mid;
+    }
+  }
```

Comment:
At the end here you'll want to return `-1`. Although maybe it makes more sense to have this method return an `Option<nat>` to make things even cleaner, and then you could return `None`.